src/Control/Monad/Extra.hs

```diff
@@ -160,6 +160,26 @@ whileM act = do
     b <- act
     when b $ whileM act
 
+-- | Keep running an operation until it becomes a 'Nothing', accumulating the
+--   monoid results inside the 'Just's as the result of the overall loop.
+whileJustM :: (Monad m, Monoid a) => m (Maybe a) -> m a
```

Comment:
Any reason to not return the more informative `m [a]` which gives all the values in order for which it was true? That way you can also simplify by dropping the `Monoid` constraint.

Reply:
I did think about that, but `Monoid` is actually more general, since a list is a free monoid anyway so this function works on any `[a]`. Apart from that, this version cover the case for `()` where the monoid instance collapses all the `()` together so the final result is `()` and not `[()]` - which you'd have to `void` on to avoid a warning.

Reply:
There is the general principle that any list could be a monoid, but not as conveniently or obviously, and with a much less discoverable API. If you want the `()` case you can use `whileM` instead, so not that interested in that. But due to a space leak, this version would actually build up a huge thunk of `() <> () <> () <> () <> ()` which would be bad. I'm still thinking list makes more sense.

Reply:
Another case, which was the one I'm using and the one I *meant* to describe, was where you want to return a value out of the inner loop at the end - then you have `a ~ Maybe b` and you end up with `Nothing <> Nothing <> Nothing <> Just b == Just b` at the end. Regarding that huge thunk, what about strict application? `(<>) $! a1 a2`. I think I originally meant it like that anyways, hence using an accumulator rather than a non-tail-call.

Reply:
Whereas - a list of `()` would remain a list, wasting space. Using the Monoid instance actually helps us collapse it during the operation, that is if we do actually write the application strictly.